### ISSUE TITLE
[REQUEST] Water flow metering reading of Kamstrup smart water meters via IR optical eye header?

### ISSUE THREAD

**Issue by Hedda** [esphome/feature-requests#1402]
**Describe the problem you have/What new integration you would like**

Like to get water flow metering from smart water meters with IR eye via hardware similar to SlimmeLezer and Home Assistant Glow.

https://www.mysensors.org/build/ir

http://jheyman.github.io/blog/pages/WirelessWaterMeter/

I believe that a such IR reader adapter could be used with newer Kamstrup water meters with IR optical heads common in Europe.

Specifically, I myself live in Sweden where I think Kamstrup MULTICA 21 / Kamstrup flowIQ 210x ultrasonic water meter is the most common model in normal private residential houses here in Scandinavia (Sweden, Norway, Denmark, Finland, and Island).

https://www.kamstrup.com/en-en/water-solutions/smart-water-meters

https://www.kamstrup.com/en-en/water-solutions/smart-water-meters/multical-21

![image](https://user-images.githubusercontent.com/6320001/132475251-d11c3a53-223b-496a-b683-7cdd5ccbd2f9.png)

![image](https://user-images.githubusercontent.com/6320001/132484030-a79e7e76-eb27-4ecd-a7ec-d34797c9cb26.png)

They currently also have Kamstrup flowIQ 2200 "advanced model with acoustic leak detection": 

![image](https://user-images.githubusercontent.com/6320001/132476561-04013531-13d9-432b-a1df-cf5d3c2985b6.png)

![image](https://user-images.githubusercontent.com/6320001/132484562-c4952c04-803a-456f-8910-c51b347c84fc.png)

and Kamstrup flowIQ 3100 "Commercial and industrial" model 

https://www.kamstrup.com/en-en/water-solutions/smart-water-meters/flowiq-2200

https://www.kamstrup.com/en-en/water-solutions/smart-water-meters/flowiq-3100

**Please describe your use case for this integration and alternatives you've tried:**

I believe many new water meters offer such two-way bidirectional infrared port (IR receiver and sender header) for communication?

https://www.mysensors.org/build/ir

http://jheyman.github.io/blog/pages/WirelessWaterMeter/

Kamstrup seems to only provide some technical documentation about all their products here but nothing about their protocols:

https://products.kamstrup.com/index.php

Kamstrup apparently call their implementation for Kamstrup Meter Protocol (KMP) and someone has reverse-engineered it here:

https://github.com/bsdphk/PyKamstrup

https://frack.nl/wiki/Stadsverwarming

https://forum.mysensors.org/topic/3525/district-heating-city-heating-stadsverwarming-mysensor-ir-sender-receiver/

https://www.domoticz.com/forum/viewtopic.php?t=11333

As I understand most of those optical probes should work with optical infrared waves be compliant with the international standard IEC 62056-21 (formerly IEC 1107) or ANSI C12.18 communications protocols for energy metering for reading utility meters?

https://en.wikipedia.org/wiki/IEC_62056

https://en.wikipedia.org/wiki/Smart_meter#Protocols

Kamstrup official (and expensive) optical read-out head tools are called 6699-099 (USB) and 699-102 (9F D-sub plug serial)

https://www.termonet.rs/pdf/Kamstrup/Optical%20Read-out%20(IR)%20Head%20-%20Data%20Sheet%20-%20English.pdf

**Additional context**

A hackerspace group in Denmark is selling these IR reading heads for about $14 US (Google translate as test written in Danish) and they wrote a note that using magnets is required as the IR port transmitter / reader eye does not activate unless it sense a magnet:

https://wiki.hal9k.dk/projects/kamstrup

PS: Off-topic but FYI; some if not all of these models of Kamstrup water meters can also be read remotely via Sigfox network RF (Radio Frequency) but that communication traffic is most of the time encrypted, at least if the water meter itself is owned and operated by a utility company or a service provider.

https://www.kamstrup.com/en-en/water-solutions/water-meter-reading

https://www.kamstrup.com/en-en/water-solutions/water-meter-reading/usb-meter-reader

If the case is that the RF radio communication is not encrypted then there look to be projects that can work with that:

https://github.com/adams-okode/kamstrup-integration

https://github.com/weetmuts/wmbusmeters

https://github.com/tobiasrask/wmbus-client

**Comment by Hedda** [comment]
This might be related the the request for SML (Smart Message Language) protocol for reading smartmeter? 

https://github.com/esphome/feature-requests/issues/1041

There is a lot collected about workings of SML (a.k.a. "D0 interface") smartmeters with IR two-way here:

https://www.msxfaq.de/sonst/bastelbude/smartmeter_d0_sml.htm

**Comment by coltju** [comment]
Check out https://github.com/golles/Home-Assistant-Sensor-MC66C

Im using this for my block heating sensor, which uses the same tech.

**Comment by evlo** [comment]
Hi, i'm looking at projects regarding this, either esphome (preferably)

OP did nice summary of the things I also found.

What i understand that IEC 62056-21 is quite widely used in this environment, and sometimes called "S0" (probably something from DIN). I even seems basically UART RS485, just trough optical IR diode(s).
_BTW I remember ages ago IR was used by laptops and cellphones to transfer data._

I think not a lot people know about this and all the projects I found so far about this are quite old/cumbersome or me or at least quite hard to follow.

So I was wondering if i understand it correctly that all need is this
https://www.aliexpress.com/item/1005003727876029.html - connect it using "dupont" cables to nodemcu32 ana i should be able to read the data if speed set to 9600 baud or slower?

**Comment by evlo** [comment]
I think the main issue is sending `/?!\r\n` / HEX: `2F3F210D0A` to poll the meter, otherwise maybe the default SML component might work.

**Comment by evlo** [comment]
Most likely what my meter spits out is not actually SML, but called IEC62056-21 or sometimes d0 - i think it goes like d0 is the port, IEC62056-21 is the definition of the IR communication, SML is the "extended" protocol and OBIS codes are what each code mean, so even though meter uses IEC62056-21, d0 and OBIS it is most likely not SML, but just IEC62056-21 compliant (which contains basic protocol, but not exactly SML - sml seems to me that it might be IEC62056-21 for multiple meters on the same bus)

This is used by more meters then just water, very similar to SML, but i think separate component is need for easiest implementation. What I wrote in my previous post about using SML component is not true

example:
![image](https://user-images.githubusercontent.com/8813196/191311406-280f66c3-2334-420f-b49a-5b6abdb17a8e.png)

**Comment by SzczepanLeon** [comment]
@evlo How to get that debug printout?

**Comment by jakobschou** [comment]
Hopefully somebody could figure out to get some cheap optical probe like this to work:
https://a.aliexpress.com/_mK4qGxM
so we dont need to solder something ourselves.

It seems that it is possible to do with other Kamstrup meters:
https://github.com/golles/ha-kamstrup_403

Though this setup is using pykamstrup (https://github.com/golles/ha-kamstrup_403/tree/main/custom_components/kamstrup_403) which I dont know if it is the right/same protocol for water meters with IEC62056-21, but maybe it could be changed to pyDLMS (https://github.com/bsdphk/PyDLMS). Here is also some code for the old IEC1107 protocol which maybe could be useful, since the original Kamstrup Optical read-out head seems to be only compliant with this protocol...

Hopefully somebody could fork kamstrup_403 and make it work with Kamstrup Multical 21/FlowIQ 2200 - I just dont have the knowledge to do so.

**Comment by evlo** [comment]
maybe take a look at

- https://github.com/evlo/esphome_ZPA-ZE312
- https://github.com/jplitza/esphome_components

using pretty much same thing

**Comment by cfeenstra1024** [comment]
If this meter uses the 'Kamstrup Meter Protocol' (KMP) we might be able to use the Kamstrup Multical40x component for ESPHome I've created recently.

If you have the hardware to connect an ESP to the optical interface, I can create a version for ESPHome that can be used to request available parameters. You can then test with your meter. If that works, I can extend the ESPHome component functionality.

I don't have this meter myself, but I do have the Kamstrup Multical 403 (district heating meter) which works well with the created component in ESPHome. I noticed that both meters mention a 'METERTOOL HCM' that can be used with the optical interface in their data sheets. This makes me think they might use the same protocol.

Some info about the component:
- [ESPHome component for Kamstrup multical 40x](https://github.com/cfeenstra1024/esphome/tree/kamstrup_mc40x_temp#readme)
- [Electronics and housing](https://github.com/cfeenstra1024/kamstrup-multical-hardware#readme) I've created for use with the Multical 40x
- [Documentation](https://deploy-preview-2551--esphome.netlify.app/components/sensor/kamstrup_mc40x.html)

**Comment by onkelbeh** [comment]
I just printed an adapter to hold an IR adapter on the meter, I am going to give it a try soon.

**Comment by glance-** [comment]
I can report that I've had success talking to a flow IQ 2200 using https://github.com/ErikErnst/kamstrup and 1200 baud.

I haven't found which variables the water meter uses yet, but I have read the time / date and serial number from the meter using a ir eye.

**Comment by jakobschou** [comment]
> I can report that I've had success talking to a flow IQ 2200 using https://github.com/ErikErnst/kamstrup and 1200 baud.
> 
> 
> 
> I haven't found which variables the water meter uses yet, but I have read the time / date and serial number from the meter using a ir eye.

What IR interface did you use?

**Comment by glance-** [comment]
Some random one from flee-bay a couple of years back. The only issue I've found is that the physical form factor of flow IQ 2200 isn't the "standard" one so i had to do some tail then error and when it finally worked apply some tape to keep the ir eye on place. "Normal" devices have a embedded metal plate and some registring features to keep the ir eye in the right place.

**Comment by copystring** [comment]
You can send `803f100100444dc00d` which the meter responds for example with `403F1000442804430000215DA390`
In my case the meter has a volume of 8.541m³ which is 215D. To get this number, take 0000215D and convert to decimal (8541). Divide it by 100, and you get the correct volume of 8.541

**Comment by copystring** [comment]
Sending `803f100100444dc00d` can be split like this:
| start byte | destination address | CID | requested var | CRC  | end byte |
| 80            | 3f10                          | 01  | 0044                | 4dc0 | 0d           |

on the receiving side:
| start byte | destination address | don't know what this is | requested var | CRC |
| 40 | 3F10 | 0044 | 280443 | 0000215D | A390 |

where 0044 stands for the current meter value

**Comment by copystring** [comment]
To calculate the checksum, ignore the start byte. In this case, take 3f10010044. Use CRC-CCITT (XModem) to calculate the CRC.

Maybe this code would work?

```
// Helper function to calculate CRC-16 (Xmodem) checksum for a buffer
function calculateCRC16(buffer) {
  var crc = 0;
  var table = [
    0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50A5, 0x60C6, 0x70E7,
    0x8108, 0x9129, 0xA14A, 0xB16B, 0xC18C, 0xD1AD, 0xE1CE, 0xF1EF
  ];
  
  for (var i = 0; i < buffer.length; i++) {
    var byte = buffer[i];
    crc = ((crc << 8) & 0xFFFF) ^ table[((crc >> 8) ^ byte) & 0xFF];
  }
  
  return crc;
}

// Example usage
var buffer = Buffer.from('803F10010044', 'hex');
var crc = calculateCRC16(buffer);

// Print the CRC checksum in hex format
console.log(crc.toString(16));
```

**Comment by aquaticus** [comment]
Hi All,

I created a component for IEC62056-21 meters. It can exchange data with utility meters. Mostly for electricity but also water, thermal and other meters.

It supports both bidirectional and unidirectional communication.
* bidirectional - you need optical interface with IR diode and phototransistor that can send and receive data. This is for modes A,B, and C.
* unidirectional - only IR phototransistor is required. This is for mode D - meter periodically sends data to the interface.

The component does not support binary encoded data (HDLC)..

Hardware required: Optical interface connected to serial port.

IEC 62056-21 term is used for multiple protocols (with the same hardware layer but different data encoding). The component supports meters that provides ASCII encoded data, something like that:
```
1-0:15.8.1(00000009999.567*kWh)
1-0:15.8.2(00000000000.000*kWh)
1-0:15.8.3(00000000000.000*kWh)
1-0:15.8.4(00000000000.000*kWh)
``` 

Detailed description of the component: https://aquaticus.info/iec62056.html

PR to ESPHome: https://github.com/esphome/esphome/pull/4388

**Comment by copystring** [comment]
Hi @aquaticus,

does this support only obis codes?

**Comment by aquaticus** [comment]
@copystring 
Not sure I understand the question. The component can only interpret information sent as OBIS codes.

**Comment by copystring** [comment]
Correct me if I'm wrong, but the way I see it, the Multical 21 does not use OBIS codes.

**Comment by aquaticus** [comment]
Just realized Multical 21 probably does not use IEC 62056-21 protocol. I was a bit mislead by the log in one of the comments. In that case the component cannot be used with Multical. And yeah, it seems Multical uses different data model than OBIS.

**Comment by glance-** [comment]
> where 0044 stands for the current meter value

I can attest for this. This works also by using readvar from https://github.com/ErikErnst/kamstrup and asking for value 68 (0x44).

280443 is decoded as 0x28 means m^3, 04 means 4 byte value and 43 means how to convert the value, aka 0x40 it's a negative exponent, and 0x03 it should be ^-3 , all according to above referenced code.

**Comment by glance-** [comment]
Btw, 74 (0x4A) gives you l/h.

**Comment by jakobschou** [comment]
@glance- So to understand - have you figured out how to connect to the meter with the IR-to-USB, and get readings from it yet that can be pushed to etc. Home Assistant?

**Comment by jakobschou** [comment]
I have found this fork that should work with the ir-to-usb with HA:

https://github.com/kristianschneider/ha-multical_21

**Comment by mroek** [comment]
@glance- I am very interested in your findings, as I also have a flowIQ 2200 in my house (installed by the municipality). The meter sends data using Wireless M-bus, but unfortulately it is encrypted, and the municipality does not want to provide me with the encryption key.

My question to you is whether you know if the data obtained through the optical eye is also going to be encrypted in my case. Do you know whether your meter also sends encrypted data wirelessly? If so, it probably means that the optical data is not encrypted.

Second question is if you have verified that you can poll the meter whenever you want, or if it imposes some limit on the optical interface to preserve battery life?

Third question is if you know whether a magnet is necessary to activate the optical interface on the meter or not. There is no metal plate for the magnet to attach to, so the official Kamstrup optical eye needs a plastic bracket that attaches to the meter to hold the optical eye in place. It could still need a magnet to activate the optical interface, though.

I guess you just used a Linux-machine to run the software you linked to (https://github.com/ErikErnst/kamstrup), but I suppose it should be possible to adapt the necessary parts of the software to run on an ESP to provide a wireless reader of sorts.

**Comment by mroek** [comment]
I have a flowIQ2200 (like @glance- ), but there is absolutely no way I can get it to talk to me. I've just DIYed a reading head (3D-printed multiple brackets for the IR leds), and I have added magnets in all kinds of orientations, but the optical interface is just completely dead.

The IR-leds I used was just something I had lying around, and I don't know their peak wavelength, but I doubt that the problem is there. If I point my IR-leds at each other, I will receive a correct echo in a serial terminal, so the circuit (even if just cobbled together on a breadbord) seems fine.

Kamstrup sells a bracket for mounting their optical eye, and due to the design of the meter, the optical eye doesn't really touch the face of the meter, instead it rests against the outer rim. I assume there are light guides in the braket to ensure a good optical connection to the meter. What this also means, is that the magnet sensor in the meter must be located inside the outer rim, and I am fairly certain this is correct. Reason is I found this text in a Kamstrup document:

**_It is possible to toggle through all views by means of the Kamstrup meter reader head, part no. 6699-
099, placed on the optical eye at the meter front. For more details on the meter reader head/optical 
eye, please see chapter 11 Tools and programs. 
The magnet needs to be removed, and applied again on the optical eye, to switch display view._**  

I found a position for the magnet where I could get the meter to change the views, so it seems natural to think that this is also the position used to turn on the optical port. However, still nothing. Very frustrating.

There is also a statement that the polarity of the magnet is important, and I did notice that the above mentioned change of display views only works with the magnet oriented with the south pole towards the meter. This is opposite to what the 62056-21 standard says (for the reading head):

**_Magnetization: axial, north pole directed towards the tariff device._**

It seems unlikely that my meter has a permanently disabled optical port, I can't find any order codes for such an option. And just to make clear, I am sending KMP telegrams, like mentioned by @copystring further up in this discussion.

**Comment by mroek** [comment]
Ok, so I finally have success! Turns out that there was indeed a problem with my receiver circuit. It wasn't sensitive enough to trigger on the IR signal from the meter.

I came up with the idea to have only my TX LED connected and sending repeated requests, and then I pointed my phone camera to the TX led of the meter. It showed something was sent, which of course prompted me to increase the sensitivity of the receiver circuit.

After that, I did get sensible data from the meter. Would be nice to have an ID for some more registers, because there are plenty of them. There's a long list in that Kamstrup document I found, but it just names them, there's no telling the address/ID.

**Comment by mroek** [comment]
I left my reader polling at 30 seconds interval, and at some point (maybe a couple of hours) the meter stopped responding. :-(

Not sure if there is some internal counter/timer to limit readouts over the optical interface to preserve battery, but it is still bad news, as it means it is not a viable method to monitor the water consumption over time. Unless of course there is an acceptable polling rate that the meter will tolerate. I also noticed that it would shut down the optical interface if it wasn't polled, even if the magnet was left in place.

Will be interesting to see if it starts responding again tomorrow, but it would be really bad design if it doesn't.

**Comment by mroek** [comment]
It did start responding again this morning, so it clearly has some battery saver function to prevent polling on the optical interface from depleting the battery. That's annoying, and I have no clue how much polling will be allowed. It does not reset at midnight either, because I tried reading it after midnight, and it was still dead. I checked the internal clock in it this morning, and it was impressively accurate (only about 1 minute off from real time).

It might have some function to count the ON-time of the TX LED (easy enough, given a fixed baud rate, so it can just count the number of transmitted 0 bits and multiply by the bit time), and stop responding after some limit has been reached. However, it must also take time into consideration somehow, and in theory it should be possible to poll continuously if the rate of polling is within the limit. I wish they could just tell us how this works, since reverse-engineering the mechanism takes a lot of time.

It might just start a countdown of 24 hours at the first poll (assuming it has been idle for more than 24 hours), and then when the ON-time limit of the LED has been reached, it will stop responding until the 24 hour countdown has elapsed. That would be a resonable implementation, and if that is indeed how it works, it makes sense to only poll for the values one needs, and not "nice-to-have" values.

**Comment by aquaticus** [comment]
@mroek Every battery powered meter applies limits for reading. The producer guarantees device will work for specific time on battery, typically 3-5 years. They compute battery lifespan by assuming max readings per day, so the power consumption is known. 
The number of readings is typically in user manual but guess you have no access to it. For example Maddalena meter I got limits readings to 4 a day.

**Comment by mroek** [comment]
@aquaticus Yes, I get that there are (and must be) limits, but it is difficult to get a handle on how this is applied for the flowIQ 2200 that I have. As mentioned, I was able to read multiple registers at a 30 second interval for at least a couple of hours continuously yesterday, but today it seems to have gone silent after just a few reads. And it also seems to deactivate the optical interface even if there is a magnet attached, so the only way to wake it again is to remove and reattach the magnet. That does not work if it has gone all silent, though.

Very frustrating to not have a proper document to refer to. I've asked Kamstrup for some info, but I have not had a reply yet.

**Comment by mroek** [comment]
The battery saving function of the flowIQ 2200 basically makes it useless. This morning I did a test where I polled it every 5 minutes, and only requesting one register. It returned 12 values (maybe a couple of more that I didn't catch in my log) before becoming unresponsive. Now, that might mean it is possible to read one value every second hour, but there's another big issue:

The optical interface also times out even if the magnet stays attached, and my tests indicates that this timeout is somewhere between 5 and 10 minutes. That means it is not possible to attach an optical head and then just poll it at a slow enough rate to avoid the polling limit (which is still unknown, just fairly low). The head must be removed and reattached to reactivate the optical interface, which is of course useless. Automating that function is of course also possible (with a servo or possibly an electromagnet), but it really makes for a clumsy solution.

A final note is that I suspect there could even be a global counter for the responses for the optical interface. That would explain why I could read a bunch of values at 30 second intervals the first time I got this working, simply because no readings had been done since the meter was installed a couple of years ago. However, now I have spent my quota, and it will just slowly refill itself.

**Comment by aquaticus** [comment]
* Maybe number of readings is limited per week or month. Reading every 30s could block the meter for a couple of days.
* Maybe the meter must be wake up by sending special sequence. Detecting magnetic field could be a one time activation.
* The reader my limit the time it is active, say to 5s, so total transmission time must be less than active time.

The above are just guessing based on other readers I played with.

**Comment by mroek** [comment]
If the magnetic field is just left in place, the meter will stop responding even if there are more "reads" available. To get responses again, the magnetic field must be removed and reapplied. However, there could exist a special (and secret) wake-up sequence, of course. It would make some sense if there was, because if service personnel wanted to read for example the logs, it would be dumb if there was no way for them to get any output because there was no quota left due to the user (i.e me) had been reading it until it stopped responding.

There is absolutely no doubt that there is a limit to the available readings, but as of yet it is impossible to know how said limit works. Could be per day, per week, it's anyone's guess, really. Or indeed transmission time. However, as we know, I was able to do a fairly large amount of readings when I first got it working, so clearly there was a number of allowed readings that had accumulated over time. That probably means it is not an amount per week, becuse if it was, then I should have been completely blocked by now. I am not, but the number of allowed readings seems to be very low by now. Probably maybe 10 per day, or something.

One observation I have made that is mildly interesting is that even if the meter has stopped responding, the optical receiver is active (thus a special wakeup sequence could exist). I know this because if you remove and reattach the magnet, you can cycle the display views. The current view will revert after a number of seconds. However, if I send something to the meter in that period, the view reverts immediately. In other words, it does read even if it does not respond.

**Comment by mdrmdrmdr** [comment]
After numerous tries with (wrong) ChatGPT suggestions and other websites, I finally could read register x0044 (total volume so far) from my Kamstrup flowIQ 2200 (got it installed just 2 weeks ago) using the code from the [Kamstrup](https://github.com/ErikErnst/kamstrup) repository on GitHub.

But I also see the "stop responding" issue. This indeed makes it quite useless for e.g. once per hour readouts. I really hope that there is some solution to get around this "feature".

Does anybody know if using the official [Kamstrup IR head](https://www.kamstrup.com/en-en/heat-solutions/meters-devices/accessories/optical-readout-head) shows the same behaviour?

**Comment by mroek** [comment]
> Does anybody know if using the official [Kamstrup IR head](https://www.kamstrup.com/en-en/heat-solutions/meters-devices/accessories/optical-readout-head) shows the same behaviour?

The optical head doesn't really do anything except convert the data stream to IR light, so I highly doubt that it would make any difference whatsoever. It is possible that the Kamstrup software has some kind of secret sequence to allow it to talk to the meter even after it has stopped responding to normal register reads, but again, that will be a feature in the software, not in the reading head.

Using the official Kamstrup software is probably the only way of finding out if there is a way of getting the meter to respond again, but the software isn't publicly available, and on top of that I suspect it isn't useable without a Kamstrup account.

@mdrmdrmdr Where in the world are you located? Maybe you could get access to the encryption key for your meter, and listen to the radio transmissions from the meter instead?

**Comment by mdrmdrmdr** [comment]
> @mdrmdrmdr Where in the world are you located? Maybe you could get access to the encryption key for your meter, and listen to the radio transmissions from the meter instead?

In wonderful Upper Bavaria :-) I doubt that I could get the encryption key. And I'd need to dive into the 868 MHz communication of the flowIQ. I already have a 868 MHz gateway for a differential temperature meter for Homegear, but don't know if this could be used to sniff the Kamstrup radio transmissions.

**Comment by mdrmdrmdr** [comment]
I also placed the question regarding the "stop responding" issue  [here](https://github.com/Apollon77/ioBroker.smartmeter/issues/333#issuecomment-1523142249).

**Comment by mroek** [comment]
@mdrmdrmdr If you can get access to the encryption key, getting the data from the 868 MHz wireless M-bus is simple enough. You'd just need a suitable radio dongle, and then you could use https://github.com/wmbusmeters/wmbusmeters/ to get the data. I did that, and I can see that my meter sends data, I just can't read it because I haven't got the encryption key.

**Comment by SzczepanLeon** [comment]
Here https://forum.arturhome.pl/t/komponent-do-esphome-umozliwiajacy-odczyt-miernikow-wm-bus-i-transmisje-wyniku-do-ha is nice discussion about wM-Bus and ESPHome.

**Comment by mdrmdrmdr** [comment]
Thanks, I'll try that. Do you have a suggestion for a good USB radio dongle? The key should be know by the local water people, right? They don't give me the impression that they would understand my question about the key...

**Comment by SzczepanLeon** [comment]
> Thanks, I'll try that. Do you have a suggestion for a good USB radio dongle?

Discussion above is about ESP + CC1101. No USB dongles, just ESPHome + external component.

**Comment by mroek** [comment]
@mdrmdrmdr Ask them to give you the KEM file for your meter (serial number of your meter is needed). They'd need to log onto the Kamstrup servers to get the file, but they may well refuse. I have had no success so far with my water meter company (the municipality), they just point to GDPR in their refusal.

**Comment by mroek** [comment]
As for a dongle, I purchased this (pretty cheap) one: https://www.aliexpress.com/item/1005002623914698.html

Seems to work OK, but again, quite useless without the key.

**Comment by mdrmdrmdr** [comment]
Your link responds with "_Sorry, the page you requested can not be found:(_"

**Comment by mroek** [comment]
@mdrmdrmdr Very odd. The link works for me, both in Chrome and FF.

**Comment by mdrmdrmdr** [comment]
Ah, a second small font comment says: "_Unfortunately, this item is currently unavailable in your location._"

**Comment by mdrmdrmdr** [comment]
I'm pretty sure that a wake-up of the IR port of my flowIQ 2200 only is possible by lifting the IR head. I don't think that there is a secret wave-up sequence... I could not find any hint for this. Sad!

Are there IR heads with a USB controllable electrical magnet? ;-)

**Comment by mroek** [comment]
> I'm pretty sure that a wake-up of the IR port of my flowIQ 2200 only is possible by lifting the IR head. I don't think that there is a secret wake-up sequence... I could not find any hint for this. Sad!
> 
> Are there IR heads with a USB controllable electrical magnet? ;-)

You are most likely correct. I am in the process of building a reader (ESP8266 based) that automates this by having a servo-controlled magnet to trigger the meter to wake up. The magnet sensor in the meter seems to be in one particular spot, so only one powerful magnet is needed, but it needs to be placed in the correct location.

Once I have this in place, I can test the max frequency of readout, which I suspect is a handful each day, tops. I also suspect (or perhaps I should say hope) that the meter counts the number of reads, so it might be good to read multiple registers in each read, instead of one read per register.

**Comment by mdrmdrmdr** [comment]
Funny, did not think that my idea is taken seriously 😅
Would a electrical magnet not be easier than a construction with a servo? Especially for those already with a IR head? Do you already know the exact required location of the magnet?

**Comment by copystring** [comment]
Electro magnets use a lot of power to work correctly and might need to be calibrated?

**Comment by mroek** [comment]
I have experimented with an electromagnet, and I'm quite sure it would be possible, but it takes very much power to generate a strong enough field to trigger the sensor. A neodymium magnet is very strong, and the one I have been using is 10x5 mm circular. It is just really simple to use a servo to move the magnet.

I'm doing this as en exercise to develop something interesting with an ESP, but I am still going to push to gain access to the encryption key. If I succeed, this contraption is of course no longer of any use, but that's OK. As I said, I am doing it to educate myself, and I do have all the necessary components lying around anyways.

**Comment by mroek** [comment]
I now have a working prototype of my reader. It does use a servo to move a magnet to wake the meter, and then performs a reading before moving the magnet out of the way again. I am currently running tests to find the limits of the polling frequency, but it will take some days until I (hopefully) have an answer.

**Comment by glance-** [comment]
There's some magic modes in the meter, evidently by the addon:
https://www.kamstrup.com/en-en/water-solutions/meters-devices/accessories/pulse-adapter

It can put the meter on some mode where it emits a "pulse" over the ir eye which then is turned into a open collector pulse output by the addon. The addon looks to be powered by a saft14500.

**Comment by mroek** [comment]
I have also connected the servo directly to a GPIO on the ESP8266, no problem there. The servo is pretty cheap, and it isn't all that precise. Good enough for this, though. I found out that just one 10x5 mm magnet wasn't quite enough, so I stacked two of them to get a stronger field to reliably trigger the meter.

The IR-circuit is also exceedingly simple. I've connected the TX LED to a GPIO through a resistor (the GPIO can sink around 20 mA, which is good enough), and the RX is a NPN phototransistor where the emitter is connected to ground, the collector to a GPIO (I'm using softwareserial, but it could just as well be an UART pin), and with a 1.5 k resistor to VCC (3.3V) to pull it up when there's no light.

Using a phototransistor makes for a simpler circuit, since it is more sensitive than a diode. I used a phototransistor called PT334-6C, which comes in a standard 5 mm package just like a LED.

**Comment by mroek** [comment]
> There's some magic modes in the meter, evidently by the addon: https://www.kamstrup.com/en-en/water-solutions/meters-devices/accessories/pulse-adapter
> 
> It can put the meter on some mode where it emits a "pulse" over the ir eye which then is turned into a open collector pulse output by the addon. The addon looks to be powered by a saft14500.

There's no documentation on that adapter. However, even if we could put the meter into that pulse mode, it still might not stay in that mode forever. It seems that the pulse interface for the flowIQ2200 is different (or it is a newer version):

![image](https://user-images.githubusercontent.com/9359176/235652218-a32211a8-d907-4e12-b32a-9a17d06e8f78.png)

The pulse interface uses the same plastic bracket as the normal Kamstrup optical reading head, as you can see (or maybe it is the same, it's just that control box that is different). The meter also stays in the pulse mode for 9 hours after removal of the pulse interface, but there's no telling if it could stay in that mode indefinitely if the interface is left on.

And if it could indeed stay in that mode, since it is a test mode, it might not actually update the normal registers, and it might also stop broadcasting the radio telegrams. That would also be a problem when the municipality does their drive-by to read the meters...

**Comment by mdrmdrmdr** [comment]
> I found out that just one 10x5 mm magnet wasn't quite enough, so I stacked two of them to get a stronger field to reliably trigger the meter. 

I have such Neodyms from my last Carrera tuning actions 😊 Can you please share the location the magnet needs to touch the meter?

**Comment by mroek** [comment]
> > I found out that just one 10x5 mm magnet wasn't quite enough, so I stacked two of them to get a stronger field to reliably trigger the meter.
> 
> I have such Neodyms from my last Carrera tuning actions 😊 Can you please share the location the magnet needs to touch the meter?

Here's a picture:

![image](https://user-images.githubusercontent.com/9359176/235751533-ee26b071-b1b9-454a-b1bf-88bcefc6011d.png)

The grey dot indicates the magnet position that I have found to work best. You can easily test this by holding the magnet to that spot for a few seconds. The display should then change mode, and by taking the magnet away and then replacing it, you can cycle the different display views. The orientation of the magnet also matters, for my meter the south pole of the magnet must face the meter. This is opposite of what the 62056-21 standard says, oddly enough.

**Comment by mdrmdrmdr** [comment]
Great, thanks a lot. I'm currently not at home, so I can't measure it, but it seems that this position is slightly outside of my 30 mm ø IR head, which would make it easier.

**Comment by mroek** [comment]
> Great, thanks a lot. I'm currently not at home, so I can't measure it, but it seems that this position is slightly outside of my 30 mm ø IR head, which would make it easier.

Different shapes of magnets would also work, as long as they are strong enough. The sensor inside probably sits at a radius of around 26 mm from the center point between the LEDs, and to me it looked like it was more or less right above the RX LED, but it could be offset slightly from the position I found. I don't think it is far off, though.

**Comment by mroek** [comment]
I have been polling the device every 30 minutes, with a read of 3 registers (in one single read command) each time. After 15 responses, it went silent after 15:00 (I got a reading at 15:00, but nothing at 15:30). The next time I had a response was 01:00 the next day, and then I had 4 subsequent responses (last one at 03:00, so 5 i in total) until it went silent again.

My current best guess is that at some time of day (in my case 01:00) the meter will add to the "reading quota", and when it has been spent, it stays silent until the next time it adds to the quota. If a meter hasn't been read for a long time, it is possible to do many reads, as the quota builds every day, no matter if it is spent or not.

I do not know whether I could have had more reads if I had asked for just one register each time, but as of now I am abandoning the project, because I have received the key to my meter. I will therefore be able to pick up the radio telegrams instead, which is of course the better approach.

**Comment by mdrmdrmdr** [comment]
What documentation do you have for the KMP protocol for flowIQ 2200? So far I did read out just one register at a time for the total volume. Btw, I'll try now a linear servo. Just ordered it. 
![IMG_9679](https://user-images.githubusercontent.com/16529884/235862590-902e7ca8-d5fb-466c-b2ec-b4ad5c94741a.jpeg)

**Comment by mroek** [comment]
There is a protocol document here:

https://github.com/nabovarme/iphonedatalogger/blob/master/KMP%20Kamstrup%20Heat%20Meter%20Protocol%20MULTICAL%20601%20and%20MULTICAL%20801.pdf

It is old, and not for the flowIQ2200, but the protocol itself is the same. Creating a request for multiple registers is simple enough, and described in there.

The registers that I have been reading, are these:

```
0x0044: "Volume register"  (m3, as shown on the display of the meter)
0x004A: "Current flow" (liters per hour)
0x00EF: "Volume" (liters)
0x03EA: "Time"
0x03EB: "Date"
0x03E9: "Serial number"
```

Good luck!

**Comment by mdrmdrmdr** [comment]
That's the doc I use as well. Did you try the things in chapter 6.3? It seem that the flowIQ stores a hourly value of register 0044 and all those values can be read with one request. I think I read about that somewhere but just can't remember were.

**Comment by mdrmdrmdr** [comment]
Now my servo successfully moves using the [pigpio C library](https://abyz.me.uk/rpi/pigpio/index.html) with PWM signals. Next is to attach the 2 magnet somehow to the servo and the servo to the IR head (without the permanent magnet). Already curious if this all works ;-)

**Comment by mdrmdrmdr** [comment]
Maybe someone here can help me on this:

This KMP protocoll request "**x80 x3f x10 x01 x00 x44 x4d xc0 x0d**" works perfectly well for a flowIQ 2200 reading (x10) one (x01) register (x00 x44) with the total volume using a USB connected IR head. I get the correct number of m3 which is also shown on the the display of the meter.

On the other hand, this reqest "**x80 x3F xA0 x01 x01 x00 x44 x0C x80 x17 x05 x09 x17 x2D x00 x00 x00 x00 xF1 x0D**" does not work. It should read from the data logger 01 (xA0 x01) one (x01) variable (x00 x44) with 12 values (x0C) and a max. respond buffer of 128 records (x80) starting at timestamp "09.05.2023 23:45:00" (x17 x05 x09 x17 x2D x00) or later.

In this case I do not get an meaningful answer and the meter responds with almost the same bytes as the request "x80 x3F xA0 x01 x01 x00 x44 x0C x1B x7F x17 x05 x09 x17 x2D x00 x00 x00 x00 xF1 x0D", which makes no sense.

Any idea what's wrong?

**Comment by mdrmdrmdr** [comment]
Please send me a quick ping to manfred@tr1.de, so that I know you read this. Then I'll delete this line again... Thanks upfront :-)

> I'd guess that you'd still be cut off by the battery saving functionality of the meter.

Well, therefore I'd like to only wake up the meter once a day and read 24 values... possibly even rarer...

**Comment by mroek** [comment]
I'll write you an email very soon. You can delete the address again now.

**Comment by mdrmdrmdr** [comment]
Just a quick update on this:

I could now almost completely analyze and breakdown the requests and responses sent from the Kamstrup LogView HCW software to and from my flowIQ 2200 to read more than just one value. After all investigations it looks even simple :-)

In short:

- The meter stores the hour values (date / time / V1 m3-value)  in a 2400 value ring-buffer (= 100 days with 24 values each)
- To find out the correct start index for the 24 hour values (00:00 ... 23:00) of the previous day (or any day within the last 100 days), I just have to read index 0 to get that date and time and then calculate the required offset to reach the date & time range requested
- Then I request in 2 steps (first 16 then 8 values; that’s how LogView does it) the 24 values

I also found out, that many basics of the KMP protocol documentation for Multical 601/801 is still valid. E.g. the calculation of date, time, integers and floats. Also the id's of the registers. And most important, the syntax for the ring-buffer index.

**Comment by cenobitedk** [comment]
@cfeenstra1024

> If this meter uses the 'Kamstrup Meter Protocol' (KMP) we might be able to use the Kamstrup Multical40x component for ESPHome I've created recently.
> 
> If you have the hardware to connect an ESP to the optical interface, I can create a version for ESPHome that can be used to request available parameters. You can then test with your meter. If that works, I can extend the ESPHome component functionality.
> 
> I don't have this meter myself, but I do have the Kamstrup Multical 403 (district heating meter) which works well with the created component in ESPHome. I noticed that both meters mention a 'METERTOOL HCM' that can be used with the optical interface in their data sheets. This makes me think they might use the same protocol.
> 
> Some info about the component:
> 
> * [ESPHome component for Kamstrup multical 40x](https://github.com/cfeenstra1024/esphome/tree/kamstrup_mc40x_temp#readme)
> * [Electronics and housing](https://github.com/cfeenstra1024/kamstrup-multical-hardware#readme) I've created for use with the Multical 40x
> * [Documentation](https://deploy-preview-2551--esphome.netlify.app/components/sensor/kamstrup_mc40x.html)

Dude! This is awesome! I believe with some modifications this can be used as generic `kmp` component to be used with all Kamstrup meters that supports KMP. 😁
We just need collect a big list of default register values that can be used as reference for available sensors, with option to overwrite the register id value to match any KMP compatible meter.

**Comment by tarmor1** [comment]
> There's some magic modes in the meter, evidently by the addon: https://www.kamstrup.com/en-en/water-solutions/meters-devices/accessories/pulse-adapter
> 

Hi all!
Another Multical 21 user here from Estonia without access, despite asking, to the AES key. The reasoning from my local water company is that as the meter get removed+calibrated+randomly replaced to new home once every 5 years, it could be that the next home it'll be fitted to is within the receiving vicinity of my wireless MBus dongle.

Be that as it may, I got the "manual" from local Kamstrup vendor for the Pulse Adapter. As it has not been posted here yet nor seems fully visible on their own website, I'll share here.

![image001](https://github.com/esphome/feature-requests/assets/106139715/81dc4052-3602-4b14-8ab4-333f1a912b68)
![image002](https://github.com/esphome/feature-requests/assets/106139715/ca6d5f0d-3929-4054-bc8b-3d50e463f59e)
![image003](https://github.com/esphome/feature-requests/assets/106139715/4e8fa593-3a7e-43c5-9259-8f9c2f33a961)
![image004](https://github.com/esphome/feature-requests/assets/106139715/005c493b-9492-42f6-9209-f32fe5663a35)


With that now, I'm not sure if I can use that one for the registries querying as posted here without the magic of automated magnet? Guess I have to buy and try.

**Comment by gbyhtn** [comment]
> I just printed an adapter to hold an IR adapter on the meter, I am going to give it a try soon.
@onkelbeh:
How are your experiences with your 3D-printed adapter.
I#m currently working with some tube-insulation cutted to fit, but it's very clumsy.
So it would be very appreciated if you can publish you 3D-modell for the Kamstrup Multical21.
THX
Ralf

**Comment by BerlinJoker** [comment]
@mdrmdrmdr & @cenobitedk & @mroek , this is some great stuff you pulled together here around reading the various registers and the device/protocol overall :) 
Has anyone of you ever successfully requested values from the long-term logging of the multical21? According to the documentation for the device ist should have storage for certain daily, monthly and even yearly values. I know those are usually not needed for continuous usage monitoring, but I'd still be interested in getting those as well.

**Comment by cenobitedk** [comment]
@BerlinJoker I don't know, but I would ask at https://github.com/htvekov/esphome_multical402 if he knows how to do that. He was successful in pulling high resolution data from registers on his meter.

**Comment by mdrmdrmdr** [comment]
@BerlinJoker: I'm pretty sure that works. In my productive setup, I'm reading daily at 01:30 am, the 24 h-values of the previous day. I trigger automatically the IR communication of the flowIQ 2200 with a neodymium magnet moved by a servo. I have to admit that I did not test it, but by just changing the requested register, it should work. I use id 0x44 "Volume register V1". To determine the proper id's, guess one needs to analyze the USB communication from and to the Kamstrup LogView HCW software with a USB sniffer while reading the week, month or year values. That's how I did it with the hour values. If you are interested, I can send you more details how I did it by mail. I won't post it here, as the approach is reverse engineering which might be a problem for Kamstrup.

**Comment by tarmor1** [comment]
There's a great solution developed: https://smartgateways.nl/en/product/kamst-ir-gateway-for-kamstrup-multical-meters/
Tried it out with my Multical 21 and got it working in a matter of minutes.

**Comment by cenobitedk** [comment]
Or get them on ebay much cheaper :) https://www.ebay.com/usr/o_electronics_o

**Comment by gbyhtn** [comment]
> Or get them on ebay much cheaper :) https://www.ebay.com/usr/o_electronics_o

Hard to believe, that you tested your recommendation. A "normal" IR-Smartreader like you recommended will not work reliable on Kamstrup water meters.

**Comment by cenobitedk** [comment]
> A "normal" IR-Smartreader like you recommended will not work reliable on Kamstrup water meters.

Based on what? It's a standard used on many industries, not a proprietary Kamstrup device.

> Hard to believe, that you tested your recommendation.

Again, based on what? I have one just like it (the one I bought is no longer listed) running for over 2 years without problems.

What's the point with your comment other then being annoying?

**Comment by mdrmdrmdr** [comment]
In my productive setup, a very cheap IR-head works perfectly well. Also true is, that my flowIQ 2200 stops responding around 10 sec. after every read (magnet attach) to save battery. I overcome this issue with a servo, which attaches a cylindrical neodymium magnet some seconds before reading the values every night. The linked more expensive Kamst-IR Gateway must have the same issue. Either you attach the IR-head manually for each read, or you have an automatic setup as I have.

![setup](https://github.com/user-attachments/assets/f36468d0-706c-43b4-8bcf-16ca5d25727a)

**Comment by cenobitedk** [comment]
> Therefore I simply don't believe that you successfully tested a simple IR-SmartMeter on a Kamstrup watermeter for a longer time.

So because you had problems you refuse to believe that others didn't? 

> The problem is that water meters work with batteries and kamstrup does not use your industrial standard alone, but somehow needs an activation via neodymium magnet or some other kind of initialization. So the reader alone might work for some, if the software supporting the kamstrup protocol, but for most combinations it will not work, or stop working after some time for batterie saving purposes. At least not out of the box.

The electronics and the software are 2 different things. If you looked at the items I linked to, you can see there's a magnet included. And the product that tarmor1 linked to is more or less identical yet with USB interface, which would not work with esphome anyway. 

However, you are correct that I didn't test it on a flowIQ 2200, I have a Multical 402 that I use the optical reader with. It also turnd off the optical interface, but after 30m (or so) without any interaction from the optical reader or the menu button.

But as you can see, others made it work just fine with a simple reader like the one I linked.

**Comment by tarmor1** [comment]
> And the product that tarmor1 linked to is more or less identical yet with USB interface, which would not work with esphome anyway.

Hey!
This [Kamst-IR Gateway](https://smartgateways.nl/en/product/kamst-ir-gateway-for-kamstrup-multical-meters/) consists of "CPU" (likely a esp32 type of thing) and IR transciever (very similar by the looks to mdrmdrmdr post). This "CPU" takes power from micro-USB. Initial setup is that it creates it's own WiFi to configure and later sits in home wifi. Interface is either MQTT or REST API. Polling frequency is configurable.
Works like a charm. I don't think it has any automatic "wake up" magnet inside.

**Comment by Inspektor95** [comment]
Hello at all!
First of all, a huge compliment to everyone who deals with the topic.
I can't get any further with my problem.
I have a Multical 21 and an IR head.
Unfortunately I can't find the right script to read the Multical 21. There are two codes for the Multical 4xx/6xx on the Tasmota website. But these don't work.
https://tasmota.github.io/docs/Smart-Meter-Interface/#kamstrup-multical-4xx-6xx-8xx

Do you have a solution for this?

Many thanks in advance, and kind regards.

**Comment by BerlinJoker** [comment]
> I have a Multical 21 and an IR head.

@Inspektor95 , as an initial check if your hardware setup works (IR head working fine, positioning on the meter correctly, meter actually switching to reading-mode), I'd recommend trying out the script from this pull request and to see if you actually get values back with your hardware: https://github.com/bsdphk/PyKamstrup/pull/6

It doesn't solve your whole problem, but it checks one of the puzzle pieces you can then rule out.

**Comment by Commifreak** [comment]
> In my productive setup, a very cheap IR-head works perfectly well. Also true is, that my flowIQ 2200 stops responding around 10 sec. after every read (magnet attach) to save battery.

Maybe not directly ESPHome related but 2200 related and maybe noteworthy:

I received an SmartGateways KamstIR reader. I had similar observations like you, but a bit different:

I placed the head on the meter and booted up the ESP. Reading was successful. I set the re-read interval to 1800 seconds (30 mins). However, I never got any further update of the readings (a later logview revealed comms timeout).

I detached the head and attached it again: worked. And it worked beyond 10 seconds and more, without re-attach the head. Either I have another version with another timeout or the IR keeps enabled as long as I communicate within XX seconds of no communication. Which would not be ideal for the battery.

Just my few observations.

**Comment by Commifreak** [comment]
> That's how I did it with the hour values. If you are interested, I can send you more details how I did it by mail. I won't post it here, as the approach is reverse engineering which might be a problem for Kamstrup.


If possible, I would help SmartGateway developer to display the remaining battery life and LEAK detection over IR head. Both infos are missing in their current firmware. Do you have some infos about these two pieces of information via IR?

**Comment by mdrmdrmdr** [comment]
**@Commifreak**, because of your additional information in your email regarding register name "**BatteryDaysLeft**", I found [this](https://documentation.kamstrup.com/docs/flowIQ_3200_EU/en-GB/Data_sheet/CONTFCBC44B9B24943AABC6D8D1C11020CC6/) document with the register name and it's ID **582** (decimal). And I just tried it:

**I send to the flowIQ 2200: 80 3F 10 01 02 46 0B E0 0D**
- 80 3F: Start Sequenz Request
- 10: Read Register
- 01: Read 1 Value
- 02 46: Register Id BatteryDaysLeft (= 582 decimal)
- 0B ED: CRC
- 0D: End Sequenz

**I get back: 40 3F 10 02 46 3E 02 00 14 20 13 0B 0D**
- 40 3F: Start Sequenz Response
- 10: Read Register
- 02 46: Register Id BatteryDaysLeft (= 582 decimal)
- 3E: Unit „Days“ (from [here](https://github.com/emontnemery/domoticz_kamstrup_mqtt/blob/development/plugin.py))
- 02: Count Bytes
- 00: Sign+Exponent = 1 (not fully sure about that byte; but I assume it’s the same logic as for a float. Just a factor for the resulting value)
- 14 20: Value = decimal 5.152
    - 5.152d x 1 (sig+exp) = **14y 52d**
- 13 0B: CRC
- 0D: End Sequenz

That would make a lot of sense with 14 years and 52 days remaining. These meters have a factory battery life of ~16 years and my flowIQ was installed around April 2023. That would be 5,812 days from the time of installation = almost 16 years. Makes sense! 😁

**Comment by peff74** [comment]
@mdrmdrmdr 
can you send me more informations about your setup and decoding of the 2200?

**Comment by mdrmdrmdr** [comment]
Sure, just drop your email here and I'll send you details about my setup and findings.
I won't post it here, as the approach is reverse engineering, which might be a problem for Kamstrup.

**Comment by peff74** [comment]
We can use peff@ji5.de.
It's a one mail account, because I want to  post my real one here ^^

**Comment by simonoff** [comment]
@mdrmdrmdr wow! You did amazing work!  Can you also send me info at oleksandr@amoniac.eu? Will be really appretiated.

**Comment by mdrmdrmdr** [comment]
@simonoff: I'll prepare an information package and send it to you...

**Comment by darkstoneathome** [comment]
@mdrmdrmdr Could you send me the package too? darkstoneathome+github@posteo.de

**Comment by mdrmdrmdr** [comment]
Tried but does not work. See my mail without attachments.
You need to give me another e-mail address where the mail server can handle more attachment data.

**Comment by mdrmdrmdr** [comment]
@darkstoneathome:
I’m honestly a bit disappointed that there was no reply at all – not even a quick “thanks, got it.”
It’s a pity that communication has become like this these days.

**Comment by simonoff** [comment]
> [@simonoff](https://github.com/simonoff) and [@darkstoneathome](https://github.com/darkstoneathome): I’m honestly a bit disappointed that there was no reply at all – not even a quick “thanks, got it.” It’s a pity that communication has become like this these days.

Hi! I'm responded to your email immediately when i got it. Maby it in spam?

**Comment by mdrmdrmdr** [comment]
> Hi! I'm responded to your email immediately when i got it. Maby it in spam? 

I'm terribly sorry. Yes, **you** replied quite after my send out. Just forgot it...

**Comment by jktjkt** [comment]
hey @mdrmdrmdr , would you mind sharing your findings about log retrieval (and also about the water meters) via KMP? I'd love to get them to https://github.com/gertvdijk/PyKMP . You can reach me at jkt@jankundrat.com .

**Comment by mdrmdrmdr** [comment]
@jktjkt, sent!

**Comment by vhdirk** [comment]
@mdrmdrmdr I would be interested in findings, too. Could you also send them to me at vhdirk@gmail.com?

**Comment by Hedda** [comment]
@mdrmdrmdr could you maybe just share your notes with the public via GitHub Gist if do not want to publish them via a GitHub Git repository?

* https://gist.github.com/

That way it will be shared to everyone and we do not have to see back and forth conversations here about trouble with e-mail which is spamming everyone here who get nothing from reading about emails being sent to personal mailboxes 🙄 

Sharing is in the spirit of open-source but let us be smart about it 😉

**Comment by jekkos** [comment]
@mdrmdrmdr i would also be interested

**Comment by mdrmdrmdr** [comment]
**@all**: I'll prepare an information-package. Just please give me a couple days...

**Comment by jktjkt** [comment]
Just FYI, I added support for log retrieval to [my fork of PyKMP](https://github.com/jktjkt/PyKMP). It's using a similar approach to what Kamstrup's own LogView application is doing. Keep in mind that the meter talks with 1200 Bd communication, so it's quite slow (as in "minutes" or "dozens of minutes" for log readout), and as usual, unless your meter is AC-powered, you will shorten the battery life of your meter with each transaction.

That Python project correctly implements different units and decimal number conversion. I've only tested it with heat meters so far, but for a water meter, the register to watch is no. 68 (or no. 239 for a "high resolution" one), based on the vendor's documentation. Looks like [esphome's KMP component](https://esphome.io/components/sensor/kamstrup_kmp/) should support that one. I can check that in maybe a month once our new water meters arrive.

**Comment by mdrmdrmdr** [comment]
I'll try to include everything I know and how I did it.

It took me months to figure out how to communicate with the flowIQ 2200 and to get my software ready for "my" production setup. I also had a regular job in between, so progress was slow ;-)

For you it should be easier to reproduce, but it will still take time.

Please note: the instructions below, especially the PHP code, are *not* a take‑and‑run solution for a flowIQ2200. No idea if the concepts are valid for other Kamstrup devices. You need to understand how it works and adapt it to your own setup.

* I had access to the KMP protocol description for the Multical 601/801 (Kamstrup_KMP_Multical_601_801.pdf; available in several home automation boards). With that I could perform the first manual readings of register 0x44 (Volume V1) using Erik Ernst's software ([https://github.com/ErikErnst/kamstrup](https://github.com/ErikErnst/kamstrup)) and a manually attached IR USB head.

* The IR sensor is connected to "/dev/ttyUSB-Water-Meter" using Linux udev rules (/dev/ttyUSBx -> /dev/ttyUSB-Water-Meter).

* The IR USB head is a cheap one from eBay ([https://www.ebay.de/itm/315022714074](https://www.ebay.de/itm/315022714074)). I removed the magnet and replaced the cable with a 7 m version (longer didn't work). Wiring diagram: USB_IR_Head_SG92R.pdf.

* I also got the original Kamstrup IR head mount from my local municipality. As far as I know, it's not available to private customers.

* Next step was reading the 2,400-value ring buffer. Unfortunately, I had no documentation for it. Kamstrup refused to provide it due to trade secrets, even after multiple attempts.

* I finally used Kamstrup's "LogView HCW" software (downloadable after creating a free account) to read the ring buffer. I recorded all sent/received bytes using Simple USB Logger ([https://www.incentivespro.com/downloads.html#usb-logger](https://www.incentivespro.com/downloads.html#usb-logger)).

* To analyze the resulting XML (20230607.xml), I wrote a PHP script (show.php) to display the communication nicely (show_output.jpg). I also created an Excel sheet (Analysis_20230606-07.xlsm) to understand the meaning of most bytes.

* The Excel contains data from June 6th and 7th, 2023, and shows the analysis for date, time, and especially V1 for the 24 data sets. Note: it uses VBA and Excel 365 "Lambda" formulas. The KMP manual was extremely helpful for decoding timestamps and float values.

* This gave me a solid understanding of the flowIQ ring buffer, its index logic, etc.

* Then I implemented everything into a production setup.

* Since the flowIQ in my case always disabled the IR interface after a few seconds (probably power saving), I use a servo with a neodymium magnet to simulate attaching the IR head. See setup.jpg / setup2.jpg.

* The servo is an SG92R. The magnet is a 7x25 mm neodymium bar mounted on a 25T servo arm, fixed with waxed cord.

* Everything is controlled by a PHP script (flowIQ_read.php + includes) with low-level C functions (flowIQ_C_functions.c/.o) running on a Raspberry Pi 4.

* Servo control uses the pigpio library ([https://abyz.me.uk/rpi/pigpio/index.html](https://abyz.me.uk/rpi/pigpio/index.html)).

* The magnet sits next to the IR head at roughly the 4 o'clock position, which turned out to be the most reliable point to activate IR communication.

* For IR head wiring, see USB_IR_Head_SG92R.pdf. I used GND for the servo PWM signal, so GND had to be connected to the braided USB shield. More than 7 m cable length didn't work.

* Handling Feb 29 and DST transitions (23- or 25-hour days) was tricky. The code was refactored many times but is now quite simple.

* The low-level functions (flowIQ_C_functions.c) are compiled on the Raspberry Pi (flowIQ_C_functions.so). The build script is included (flowIQ_C_functions_build.sh).

* "automation_std.php" contains general functions for my home automation. "flowIQ_read.php" mainly uses the "msg" function from there. I removed passwords for my push service.

* "flowIQ_read.php" is the main program. It runs nightly via crontab, reads the previous day's 24 (or 23/25) hourly V1 values from 01:00 to 00:00, and writes them to a data file (1 day: dat_output.txt) and log file (1 day: log_output.txt).

* The script does:

  * Load low-level functions
  * Move magnet to the flowIQ and wait
  * Initialize IR head
  * Read meter time, flags, ring buffer boundaries
  * Calculate required ring buffer start index
  * Read 24 (or 23/25) V1 values
  * Move magnet back
  * Write values to the data file

It sounds complicated, and it is at first. It took me 3-4 months to analyze everything. But if you really dive into it, it's understandable.

You'll definitely need to adjust some things, like paths.

All mentioned files are in the attached [flowIQ_read.zip](https://github.com/user-attachments/files/26479762/flowIQ_read.zip). Password is "flow!IQ!" (without quotes).

If you have questions, just ask :-)

**Comment by vhdirk** [comment]
Thanks a lot!
